nifi-nar-bundles/nifi-data-generation-bundle/nifi-data-generation-processors/src/main/java/org/apache/nifi/processors/generation/GenerateRecord.java

```diff
@@ -0,0 +1,209 @@
+        _rels.add(REL_SUCCESS);
+        _rels.add(REL_FAILURE);
+        _rels.add(REL_ORIGINAL);
+
+        PROPS = Collections.unmodifiableList(_temp);
+        RELS  = Collections.unmodifiableSet(_rels);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPS;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELS;
+    }
+
+    private volatile RecordSetWriterFactory writerFactory;
+
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
+        writerFactory = context.getProperty(WRITER).asControllerService(RecordSetWriterFactory.class);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = null;
+        if (context.hasIncomingConnection()) {
+            input = session.get();
+
+            if (input == null && context.hasNonLoopConnection()) {
+                return;
+            }
+        }
+
+        final int limit;
+        final boolean fixed = context.getProperty(FIXED_SIZE).asBoolean();
+        if (fixed) {
+            limit = context.getProperty(LIMIT).evaluateAttributeExpressions(input).asInteger();
+        } else {
+            int ceiling = context.getProperty(LIMIT).evaluateAttributeExpressions(input).asInteger();
+            limit = new Random().nextInt(ceiling);
+        }
+
+        final Generator generator;
+        final RecordSchema schema;
+
+        try {
+            if (!context.getProperty(SCHEMA).isSet() && input != null) {
+                schema = writerFactory.getSchema(input.getAttributes(), null);
+                String text = schema.getSchemaText().get();
+                generator = new Generator(text, new Random());
+            } else if (!context.getProperty(SCHEMA).isSet() && input == null) {
+                throw new ProcessException("When there is no incoming connection, a avro schema must be set " +
+                        "in the Schema configuration property for this processor.");
+            } else {
+                Schema parsed = new Schema.Parser().parse(context.getProperty(SCHEMA).getValue());
+                schema = AvroTypeUtil.createSchema(parsed);
+                generator = new Generator(parsed, new Random());
+            }
+
+            List<Record> records = new ArrayList<>();
+            for (int x = 0; x < limit; x++) {
+                GenericData.Record o = (GenericData.Record) generator.generate();
+                Map y = AvroTypeUtil.convertAvroRecordToMap(o, schema);
+                records.add(new MapRecord(schema, y));
+            }
+
+            FlowFile out = session.create(input);
+            out = session.write(out, outputStream -> {
+                try {
+                    RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, outputStream);
+                    writer.beginRecordSet();
```

Comment:
If you create a `ListRecordSet` from `records`, you can do a single `writer.write()` and the default implementation will do the loop for you.